Add ShellCheck to `check.sh`

danielrainer opened the pull request:
ShellCheck can detect some common issues in shell scripts (not in fish scripts though). Adding it to `check.sh` means such issues are noticed and addressed. Disabling specific lints for a line can be done with a comment as demonstrated in the changes in `check.sh`.

Some context: https://github.com/fish-shell/fish-shell/pull/11600#discussion_r2160512475

## TODOs:
- [ ] Do we want a more noticeable warning when ShellCheck is skipped?
- [ ] Should we document ShellCheck as a recommended dev dependency somewhere?

danielrainer commented on .github/workflows/main.yml:
```diff
@@ -19,7 +19,7 @@ jobs:
     - name: Install deps
       run: |
-        sudo apt install gettext libpcre2-dev python3-pexpect python3-sphinx tmux
+        sudo apt install gettext libpcre2-dev python3-pexpect python3-sphinx shellcheck tmux
```
> The current approach is "duplicate", so it wouldn't really hurt doing that for shellcheck for now also.

Sure, that's reasonable for now. Autogeneration seems like it would be quite a bit of work to implement. Adding flags to `build_tools/check.sh` is probably the more realistic future approach, although it does not solve the problem of having to remember changing all users when a check is added/removed.

Commit:
Fix shellcheck warnings

Commit:
Add shellcheck to `check.sh`

Only run it when lints are enabled. Fail if shellcheck is not installed.

Commit:
Add shellcheck to CI

danielrainer commented on build_tools/check.sh:
```diff
@@ -24,3 +24,9 @@ cargo test --no-default-features --workspace --all-targets
 FISH_GETTEXT_EXTRACTION_FILE=$template_file "$repo_root/tests/test_driver.py" "$build_dir"
+if command -v shellcheck; then
+  # shellcheck disable=SC2046
+  shellcheck $(find . -name '*.sh')
```
OTOH, the current, naive approach of looking for files with the `.sh` extension misses some scripts, i.e. the ones in `build_tools/osx_package_scripts` and `debian`. (And maybe more that I've missed?)

We could use ripgrep for detecting files with a shebang indicating a non-fish shell script, e.g.:
```fish
rg --stop-on-nonmatch --files-with-matches '^#!.*[^i]sh'
``` 
Using this directly in `check.sh` would require having ripgrep installed. Alternatively, we could have a file which stores all paths we want checked by ShellCheck. This file could be autogenerated by the command above (or something else, depending on our needs). Then we could regenerate it in CI to fail on changes, which would automatically notify us when anything changes, and lets us decide whether to add the path so it gets shellchecked, or modify the autogeneration to exclude it. If this only runs in CI then checks can be added or removed locally be modifying the file containing the paths to check.

We don't have to use ripgrep of course, I just used it because it makes it simple to get what I wanted in a single command, including recursive directory traversal while adhering to gitignore files.

krobelus commented:
> > all tools that use Git to detect the project root
> 
> I think it's not limited to root detection, pretty much all tools using git will break in a non-colocated jj workspace, right?

Yeah; for now I'm only using colocated repos (except for some deployment scenarios),
and do everything in a single workspace.
[In future](https://github.com/jj-vcs/jj/pull/4644) colocation will also work for secondary workspaces, so each workspace will be a valid Git worktree.

But it's still nice to have the option to swap out the VCS without breaking unrelated things (like detecting project root).

Commit:
Add tooling for running ShellCheck

ShellCheck needs to be told which files to check. Identifying these files
automatically is nontrivial. File extensions are insufficient, since we have
several shell scripts which do not have an extension. Instead, we look for
relevant shebangs in the first line of all files in the repo and consider these
files for checking.

Because there might be files which should not be checked for whatever reason, we
add a way to exclude them by putting their paths into a dedicated file. At the
moment, no files are excluded.

Add a script for generating the resulting list of paths to check. The script can
also be used to check if the list is up-to-date.

danielrainer commented:
I implemented a list tracking which files should be shellchecked, plus a script for updating the list and checking if it is outdated, plus a way to locally ignore files. It works well locally, but for some reason `rg` fails to run correctly in CI. I don't know what's going on there. See https://github.com/danielrainer/fish-shell/actions/runs/16997092097/job/48190296278 for some tests.

krobelus commented on .github/workflows/shellcheck.yml:
```diff
@@ -0,0 +1,29 @@
+name: ShellCheck
```
I thought that this had become surprisingly complicated and that there must be a simpler solution.
But looking at it now, I guess it's fine modulo my comments, so if you're okay with that, could you rebase?

krobelus commented on .github/workflows/shellcheck.yml:
```diff
@@ -0,0 +1,29 @@
+  contents: read
+
+jobs:
+  file-list-up-to-date:
```
can we add this to the "lint" workflow in .github/workflows/lint.yml instead?

krobelus commented on build_tools/check.sh:
```diff
@@ -41,6 +42,23 @@ template_file=$(mktemp)
+    if command -v rg; then
+        echo 'Checking if list of files to shellcheck is up to date.'
+        echo 'If this fails, you can update the list automatically using'
+        echo "$shellcheck_dir/check_for_file_updates.sh --update"
```
I guess we could also tweak the diff output a little bit and then suggest

	check_for_file_updates.sh | git apply

instead (or "patch -p1" but "git apply" should be more accessible).

krobelus commented on build_tools/check.sh:
```diff
@@ -41,6 +42,23 @@ template_file=$(mktemp)
+        "$shellcheck_dir/check.sh"
+    else
+        echo 'Error: ShellCheck is not available.'
+        exit 1
```
> Do we want a more noticeable warning when ShellCheck is skipped?
 > Should we document ShellCheck as a recommended dev dependency somewhere?

failing `build_tools/check.sh` like we do here seems good enough

krobelus commented on build_tools/shellcheck/check_for_file_updates.sh:
```diff
@@ -0,0 +1,95 @@
+# Make sure that the file does not have any other content or empty lines.
+# Also ensure that `excluded.txt` is not empty.
+# The file is intended for locally ignoring scripts not committed in the repo, which is why
+# `excluded.txt` is already gitignored.
```
I don't think we should invent our own excluded file, `.git/info/exclude` already works, see my other comment.
No need to try to prevent power users from shooting themselves in the foot (which is super unlikely anyway here).
So I think we can get rid of the excluded logic?

krobelus commented on build_tools/shellcheck/check_for_file_updates.sh:
```diff
@@ -0,0 +1,95 @@
+# If a file which is committed into the repo should not be checked, disable the checks via
+# annotations in that file, either for each lint individually or by adding
+# # shellcheck disable=all
+# at the start of the file (but after the sheband).
```
sheband -> shebang

krobelus commented on build_tools/shellcheck/check_for_file_updates.sh:
```diff
@@ -0,0 +1,95 @@
+cleanup () {
+    # shellcheck disable=2317
+    if [ -n "$tmp_file" ] && [ -e "$tmp_file" ]; then
+        rm "$tmp_file"
```
due to the traps below, this could run twice, so let's make sure it's actually idempotent
(else we might remove a tmp file created by another process).
maybe define a `cleanup_done=false` variable.

krobelus commented on build_tools/shellcheck/check_for_file_updates.sh:
```diff
@@ -0,0 +1,95 @@
+# non-colocated jj repos, ripgrep cannot identify the repo root, resulting in .gitignore files not
+# being taken into account.
+# The `--no-require-git` flag prevents this, at the cost of also respecting .gitignore files outside
+# of the repo on in ancestor directories.
```
typo: on -> or

 > in ancestor directories

hmm, can't we prevent this by telling rg not to look at .gitignore files above $repo_root ?
I can't immediately see how but it's not a big deal.

We could also drop `--no-require-git` and deal with this when it becomes an actual problem (which is not anytime soon, maybe never).

krobelus commented on build_tools/shellcheck/check_for_file_updates.sh:
```diff
@@ -0,0 +1,95 @@
+#
+# By default, ripgrep also considers rules in .git/info/exclude, which only exist locally.
+# This script should behave the same way everywhere, so we do not want to take local configuration
+# into account, which is why we disable this feature with `--no-ignore-exclude`.
```
I think this is needlessly aggressive.
I don't think we should lint scripts that have been explicitly ignored by the user.
It's true that things like clippy do but I often write temporary shell scripts and linters would get in the way,
and even worse, the check_for_file_updates script would try to add them.

It's weird to respect some gitignore rules but not others.
Let's remove the --no-ignore-exclude flag? The important thing is that anything that will be tracked in Git will be checked, which we'll still have.

krobelus commented on build_tools/shellcheck/check_for_file_updates.sh:
```diff
@@ -0,0 +1,95 @@
+#
+# Finally, the output is sorted, because ripgrep's output is unsorted.
+# Setting the locale is important, because the behavior of sort is locale-dependent.
+rg --no-require-git --no-ignore-exclude --stop-on-nonmatch --files-with-matches '^#!.*[^i]sh' \
```
maybe better to use negative lookahead `(?!fish)sh` (maybe need `rg --pcre2`) but no big deal.

danielrainer replied:
I quickly benchmarked these two commands:
```sh
rg --stop-on-nonmatch --files-with-matches '^#!.*[^i]sh'
rg --pcre2 --stop-on-nonmatch --files-with-matches '^#!.*(?!fi)sh'
```
The former takes about 15 ms, the latter 32 ms on my system, so I think we should use the former approach unless you are concerned about false positives.

krobelus replied:
> `'^#!.*(?!fi)sh'`

my bad, this is actually trivially true. Needs negative lookbehind i..e `(?<!fi)sh`

danielrainer replied:
Doesn't make much of a difference for performance. For the new Rust implementation I used the `pcre2` crate we already use, although `ignore` already introduces some `regex` dependencies, so I think we could use that as well. Performance-wise, there is not much of a difference between the two with `pcre2`, at least for debug builds of the whole xtask, which is what I think matters in practice, so we could use whichever you prefer.
